libs/language-server/src/lib/ast/expressions/operator-evaluator.ts

```diff
@@ -0,0 +1,169 @@
+//
+// SPDX-License-Identifier: AGPL-3.0-only
+
+import { strict as assert } from 'assert';
+
+import { ValidationContext } from '../../validation/validation-context';
+import { BinaryExpression, UnaryExpression } from '../generated/ast';
+// eslint-disable-next-line import/no-cycle
+import {
+  BinaryExpressionOperator,
+  UnaryExpressionOperator,
+} from '../model-util';
+
+import { EvaluationStrategy, evaluateExpression } from './evaluation';
+
+export type OperandValue = boolean | number | string;
+
+export type OperandValueTypeguard<T extends OperandValue> = (
+  value: OperandValue,
+) => value is T;
+
+export const OPERAND_VALUE_TYPEGUARD: OperandValueTypeguard<OperandValue> = (
+  value: OperandValue,
+): value is OperandValue => {
+  return true;
+};
+
+export const NUMBER_TYPEGUARD: OperandValueTypeguard<number> = (
+  value: OperandValue,
+): value is number => {
+  return typeof value === 'number';
+};
+
+export const BOOLEAN_TYPEGUARD: OperandValueTypeguard<boolean> = (
+  value: OperandValue,
+): value is boolean => {
+  return typeof value === 'boolean';
+};
+
+export interface OperatorEvaluator<
+  E extends UnaryExpression | BinaryExpression,
+> {
+  readonly operator: E['operator'];
+
+  evaluate(
+    expression: E,
+    strategy: EvaluationStrategy,
+    context: ValidationContext | undefined,
+  ): OperandValue | undefined;
+}
+
+export abstract class DefaultUnaryOperatorEvaluator<
+  O extends OperandValue,
+  T extends OperandValue,
+> implements OperatorEvaluator<UnaryExpression>
+{
+  constructor(
+    public readonly operator: UnaryExpressionOperator,
+    private readonly operandValueTypeguard: OperandValueTypeguard<O>,
+  ) {}
+
+  protected abstract doEvaluate(
+    operandValue: O,
+    expression: UnaryExpression,
+    context: ValidationContext | undefined,
+  ): T | undefined;
+
+  evaluate(
+    expression: UnaryExpression,
+    strategy: EvaluationStrategy,
+    context: ValidationContext | undefined,
+  ): T | undefined {
+    assert(expression.operator === this.operator);
+    const operandValue = evaluateExpression(
+      expression.expression,
+      strategy,
+      context,
+    );
+    if (operandValue === undefined) {
+      return undefined;
+    }
+
+    assert(this.operandValueTypeguard(operandValue));
+
+    return this.doEvaluate(operandValue, expression, context);
+  }
+}
+
+export abstract class DefaultBinaryOperatorEvaluator<
+  L extends OperandValue,
+  R extends OperandValue,
+  T extends OperandValue,
+> implements OperatorEvaluator<BinaryExpression>
+{
+  constructor(
+    public readonly operator: BinaryExpressionOperator,
+    private readonly leftValueTypeguard: OperandValueTypeguard<L>,
+    private readonly rightValueTypeguard: OperandValueTypeguard<R>,
+  ) {}
+
+  protected abstract doEvaluate(
+    leftValue: L,
+    rightValue: R,
+    expression: BinaryExpression,
+    context: ValidationContext | undefined,
+  ): T | undefined;
```

Comment:
If I understand it correctly, we return `undefined` on failure. (a) We should document that, and (b) we should have it on the radar to change it as soon as we have expressions that can evaluate to undefined (if that ever should be the case).